qustion 4:

, max_ 11x0+13x1+9x2+15x3+**12x4**

6.5x0+8x1+10.5x2+11.75x3+14.5x4<=78,
5.75x0+4.5x1+3.75x2+4.85x3+3.45x4<=84,,
x0,x1,x2,x3,x4, are number of  drone  model
x0,x1,x2,x3,x4 are all integer becuase the are drone model 

In [257]:
!pip install -q pyomo

In [258]:
from pyomo.environ import * 

In [259]:
import numpy as np

In [260]:
import pandas as pd

In [261]:
data_csvfile = pd.read_csv('lab4_ex3.csv')

In [262]:
data_csvfile

Drone Model  Carrying Capacity(kgs)  Cost(in Crores of rupees)  \
0       SHREY                      11                       6.50   
1      DHARAN                      13                       8.00   
2       SATVA                       9                      10.50   
3       ADHYA                      15                      11.75   
4      PRAVAT                      12                      14.50   

   Parking Space(sq.m.)  
0                  5.75  
1                  4.50  
2                  3.75  
3                  4.85  
4                  3.45

In [263]:
print (type(data_csvfile))

<class 'pandas.core.frame.DataFrame'>


In [264]:
data_csvfile['Drone Model']

0     SHREY
1    DHARAN
2     SATVA
3     ADHYA
4    PRAVAT
Name: Drone Model, dtype: object

In [265]:
data_csvfile['Carrying Capacity(kgs)']

0    11
1    13
2     9
3    15
4    12
Name: Carrying Capacity(kgs), dtype: int64

In [266]:
data_csvfile['Cost(in Crores of rupees)']

0     6.50
1     8.00
2    10.50
3    11.75
4    14.50
Name: Cost(in Crores of rupees), dtype: float64

In [267]:
data_csvfile['Parking Space(sq.m.)']

0    5.75
1    4.50
2    3.75
3    4.85
4    3.45
Name: Parking Space(sq.m.), dtype: float64

In [268]:
data_csvfile.index

RangeIndex(start=0, stop=5, step=1)

In [269]:
len(data_csvfile.index)

5

In [270]:
model = ConcreteModel()

In [271]:
N = len(data_csvfile.index)

In [272]:
M = 2

In [273]:
col_indices = range(N)
print(col_indices )

range(0, 5)


In [274]:
model.x = Var(col_indices)

In [275]:
model.objective = Objective(expr=summation(data_csvfile['Carrying Capacity(kgs)'], model.x),sense=maximize)

In [276]:
model.constraints = ConstraintList()

In [277]:
model.constraints.add(expr=6.5* model.x[0] +8* model.x[1] +10.5* model.x[2]+11.75*model.x[3]+14.5*model.x[4]<= 78  )
model.constraints.add(expr= 5.75*model.x[0] +4.5* model.x[1] + 3.75* model.x[2] +4.85*model.x[3]+3.45*model.x[4]<= 84  )


In [278]:
lower_bound = 0
upper_bound = np.inf

In [279]:

for j in col_indices:
  model.x[j].setlb(lower_bound)
  model.x[j].setub(upper_bound)

In [280]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True :  Reals
          1 :     0 :  None :  None : False :  True :  Reals
          2 :     0 :  None :  None : False :  True :  Reals
          3 :     0 :  None :  None : False :  True :  Reals
          4 :     0 :  None :  None : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
    constraints : Size=2, I

In [281]:
!apt-get install -y -qq coinor-cbc

In [282]:
opt_cbc = SolverFactory('cbc')

In [283]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [284]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('  Drone Model ',data_csvfile['Drone Model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  132.0

Decision Variables
  Drone Model  SHREY 12.0
  Drone Model  DHARAN 0.0
  Drone Model  SATVA 0.0
  Drone Model  ADHYA 0.0
  Drone Model  PRAVAT 0.0

Constraints
constraints : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


when dicision variable are real number the Objective =  132.0

Decision Variables
  Drone Model  SHREY 12.0
  Drone Model  DHARAN 0.0
  Drone Model  SATVA 0.0
  Drone Model  ADHYA 0.0
  Drone Model  PRAVAT 0.0


# now we dicision variable are non negative integers

In [285]:
for i in col_indices:
  model.x[i].domain=NonNegativeIntegers

In [286]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  12.0 :  None : False : False : NonNegativeIntegers
          1 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :   0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3]

In [287]:
result=opt_cbc.solve(model)
print('solveer status:', result.solver.status)
print('solver termination condition :',result.solver.termination_condition)

solveer status: ok
solver termination condition : optimal


In [288]:
print('\nObjective=',model.objective())
print('\dicision variable')
for i in col_indices:
    print('Num slots for  Drone Model ',data_csvfile['Drone Model'][i], model.x[i].value)
print('\nConstraints')
model.constraints.display()


Objective= 132.0
\dicision variable
Num slots for  Drone Model  SHREY 12.0
Num slots for  Drone Model  DHARAN 0.0
Num slots for  Drone Model  SATVA 0.0
Num slots for  Drone Model  ADHYA 0.0
Num slots for  Drone Model  PRAVAT 0.0

Constraints
constraints : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


when dicision variable are non negative integer then the bjective= 132.0


dicision variable
Num slots for  Drone Model  SHREY 12
Num slots for  Drone Model  DHARAN 0.0
Num slots for  Drone Model  SATVA 0.0
Num slots for  Drone Model  ADHYA 0.0
Num slots for  Drone Model  PRAVAT 0.0


In [289]:
for i in col_indices:
  model.x[i].domain=Reals

In [290]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  12.0 :  None : False : False :  Reals
          1 :     0 :   0.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
    constraints : Size=2, I

In [291]:
result=opt_cbc.solve(model)
print('solveer status:', result.solver.status)
print('solver termination condition :',result.solver.termination_condition)

solveer status: ok
solver termination condition : optimal


In [292]:
print('\nObjective=',model.objective())
print('\dicision variable')
for i in col_indices:
    print('Num slots for  Drone Model ',data_csvfile['Drone Model'][i], model.x[i].value)
print('\nConstraints')
model.constraints.display()


Objective= 132.0
\dicision variable
Num slots for  Drone Model  SHREY 12.0
Num slots for  Drone Model  DHARAN 0.0
Num slots for  Drone Model  SATVA 0.0
Num slots for  Drone Model  ADHYA 0.0
Num slots for  Drone Model  PRAVAT 0.0

Constraints
constraints : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


Objective= 132.0
\dicision variable
Num slots for  Drone Model  SHREY 12.0
Num slots for  Drone Model  DHARAN 0.0
Num slots for  Drone Model  SATVA 0.0
Num slots for  Drone Model  ADHYA 0.0
Num slots for  Drone Model  PRAVAT 0.0


Q8-- ans_--
in this case using real number and integer the objetive value are  same that reason there no need for  rounding 
drone model  Carrying Capacity is integer that reson when using integer and using real number are same 



Q9``` first we take dicision variable are real number (LP)

6.5x0+8x1+10.5x2+11.75x3+14.5x4<=76,
5.75x0+4.5x1+3.75x2+4.85x3+3.45x4<=87

In [293]:
model.constraints[1].deactivate()
model.constraints[2].deactivate()

In [294]:
model.constraints.add(expr=6.5* model.x[0] +8* model.x[1] +10.5* model.x[2]+11.75*model.x[3]+14.5*model.x[4]<= 76  )
model.constraints.add(expr= 5.75*model.x[0] +4.5* model.x[1] + 3.75* model.x[2] +4.85*model.x[3]+3.45*model.x[4]<= 87  )


In [295]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('solver termination condition:',result.solver.termination_condition)
print('solver time:,result.solver.time')


Problem: 
- Name: unknown
  Lower bound: 128.6153846
  Upper bound: 128.6153846
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 6
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.021434783935546875
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
solver termination condition: optimal
solver time:,result.solver.time


In [296]:
#display solution
print('\nobjetive ',model.objective())
print('\ndicision variable')
for i in col_indices:
    print('  Drone Model ',data_csvfile['Drone Model'][i], model.x[i].value)
print('\nConstraints')
model.constraints.display()


objetive  128.615388

dicision variable
  Drone Model  SHREY 11.692308
  Drone Model  DHARAN 0.0
  Drone Model  SATVA 0.0
  Drone Model  ADHYA 0.0
  Drone Model  PRAVAT 0.0

Constraints
constraints : Size=4
    Key : Lower : Body              : Upper
      3 :  None : 76.00000200000001 :  76.0
      4 :  None :         67.230771 :  87.0


9-- now take dicision variable are non negtive integer (MILF)

In [297]:
for i in col_indices:
  model.x[i].domain=NonNegativeIntegers

In [298]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.692308 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0

In [299]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('solver termination condition:',result.solver.termination_condition)
print('solver time:,result.solver.time')


Problem: 
- Name: unknown
  Lower bound: 127.0
  Upper bound: 127.0
  Number of objectives: 1
  Number of constraints: 2
  Number of variables: 5
  Number of binary variables: 0
  Number of integer variables: 5
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.01
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.03438901901245117
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
solver termination condition: optimal
solver time:,result.solver.time


In [300]:
#display solution
print('\nobjetive ',model.objective())
print('\ndicision variable')
for i in col_indices:
    print('  Drone Model ',data_csvfile['Drone Model'][i], model.x[i].value)
print('\nConstraints')
model.constraints.display()


objetive  127.0

dicision variable
  Drone Model  SHREY 8.0
  Drone Model  DHARAN 3.0
  Drone Model  SATVA 0.0
  Drone Model  ADHYA 0.0
  Drone Model  PRAVAT 0.0

Constraints
constraints : Size=4
    Key : Lower : Body : Upper
      3 :  None : 76.0 :  76.0
      4 :  None : 59.5 :  87.0


9 ans-- 
dicision variable are real number then
objetive  128.615388

dicision variable
  Drone Model  SHREY 11.692308
  Drone Model  DHARAN 0.0
  Drone Model  SATVA 0.0
  Drone Model  ADHYA 0.0
  Drone Model  PRAVAT 0.0

when using dicision variable are integer
  objetive  127.0

dicision variable
  Drone Model  SHREY 8.0
  Drone Model  DHARAN 3.0
  Drone Model  SATVA 0.0
  Drone Model  ADHYA 0.0
  Drone Model  PRAVAT 0.0
  budget is decreased to 76 crores of
rupees and the parking space is increased to 87 units that reason optimal solution change

Q10_
frist take real number(LP)
6.5x0+8x1+10.5x2+11.75x3+14.5x4<=74
5.75x0+4.5x1+3.75x2+4.85x3+3.45x4<=88

In [301]:
model.constraints[3].deactivate()
model.constraints[4].deactivate()

In [302]:
for i in col_indices:
  model.x[i].domain=Reals

In [303]:
model.constraints.add(expr=6.5* model.x[0] +8* model.x[1] +10.5* model.x[2]+11.75*model.x[3]+14.5*model.x[4]<= 74)
model.constraints.add(expr= 5.75*model.x[0] +4.5* model.x[1] + 3.75* model.x[2] +4.85*model.x[3]+3.45*model.x[4]<= 88  )


In [304]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   8.0 :  None : False : False :  Reals
          1 :     0 :   3.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
    constraints

In [305]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('solver termination condition:',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 125.2307692
  Upper bound: 125.2307692
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 6
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.02504444122314453
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
solver termination condition: optimal


In [306]:
#display solution
print('\nobjective ',model.objective())
for i in col_indices:
  print('\n Dorne model',data_csvfile['Drone Model'][i],model.x[i].value)

print('\nconstraint')
model.constraints.display()


objective  125.230765

 Dorne model SHREY 11.384615

 Dorne model DHARAN 0.0

 Dorne model SATVA 0.0

 Dorne model ADHYA 0.0

 Dorne model PRAVAT 0.0

constraint
constraints : Size=6
    Key : Lower : Body        : Upper
      5 :  None :  73.9999975 :  74.0
      6 :  None : 65.46153625 :  88.0


now we take dicision variable are non negetive integer(MILF)

In [307]:
for i in col_indices:
  model.x[i].domain=NonNegativeIntegers

In [308]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.384615 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 

In [309]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('solver termination condition:',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 123.0
  Upper bound: 123.0
  Number of objectives: 1
  Number of constraints: 2
  Number of variables: 5
  Number of binary variables: 0
  Number of integer variables: 5
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.01
  Wallclock time: 0.01
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 13
  Error rc: 0
  Time: 0.03294062614440918
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
solver termination condition: optimal


In [310]:
#display solution
print('\nobjective ',model.objective())
for i in col_indices:
  print('\n Dorne model',data_csvfile['Drone Model'][i],model.x[i].value)

print('\nconstraint')
model.constraints.display()


objective  123.0

 Dorne model SHREY 10.0

 Dorne model DHARAN 1.0

 Dorne model SATVA 0.0

 Dorne model ADHYA 0.0

 Dorne model PRAVAT 0.0

constraint
constraints : Size=6
    Key : Lower : Body : Upper
      5 :  None : 73.0 :  74.0
      6 :  None : 62.0 :  88.0


 10 ans-
 using the dicision variable are real number
 objective  125.230765

 Dorne model SHREY 11.384615

 Dorne model DHARAN 0.0

 Dorne model SATVA 0.0

 Dorne model ADHYA 0.0

 Dorne model PRAVAT 0.0
 # when using integer
 objective  123.0

 Dorne model SHREY 10.0

 Dorne model DHARAN 1.0

 Dorne model SATVA 0.0

 Dorne model ADHYA 0.0

 Dorne model PRAVAT 0.0





10--
frist we take real number(LP)
6.5x0+8x1+10.5x2+11.75x3+14.5x4<=72
5.75x0+4.5x1+3.75x2+4.85x3+3.45x4<=89

In [311]:
model.constraints[5].deactivate()
model.constraints[6].deactivate()

In [312]:
for i in col_indices:
  model.x[i].domain=Reals

In [313]:
model.constraints.add(expr=6.5* model.x[0] +8* model.x[1] +10.5* model.x[2]+11.75*model.x[3]+14.5*model.x[4]<= 72)
model.constraints.add(expr= 5.75*model.x[0] +4.5* model.x[1] + 3.75* model.x[2] +4.85*model.x[3]+3.45*model.x[4]<= 89 )


In [314]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  10.0 :  None : False : False :  Reals
          1 :     0 :   1.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
    const

In [315]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('termination condition',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 121.8461538
  Upper bound: 121.8461538
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 6
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.024139404296875
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
termination condition optimal


In [316]:
print('\n objective',model.objective())
for i in col_indices:
  print('\n Drone Model',data_csvfile['Drone Model'][i], model.x[i].value)

print('\nconstraint')
model.constraints.display()


 objective 121.84615300000002

 Drone Model SHREY 11.076923

 Drone Model DHARAN 0.0

 Drone Model SATVA 0.0

 Drone Model ADHYA 0.0

 Drone Model PRAVAT 0.0

constraint
constraints : Size=8
    Key : Lower : Body               : Upper
      7 :  None :         71.9999995 :  72.0
      8 :  None : 63.692307250000006 :  89.0


now we take non negtive integer(MILP)

In [317]:
for i in col_indices:
  model.x[i].domain=NonNegativeIntegers

In [318]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.076923 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maxim

In [319]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('termination condition',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 121.0
  Upper bound: 121.0
  Number of objectives: 1
  Number of constraints: 2
  Number of variables: 5
  Number of binary variables: 0
  Number of integer variables: 5
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.03455066680908203
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
termination condition optimal


In [320]:
print('\n objective',model.objective())
for i in col_indices:
  print('\n Drone Model',data_csvfile['Drone Model'][i], model.x[i].value)

print('\nconstraint')
model.constraints.display()


 objective 121.0

 Drone Model SHREY 11.0

 Drone Model DHARAN 0.0

 Drone Model SATVA 0.0

 Drone Model ADHYA 0.0

 Drone Model PRAVAT 0.0

constraint
constraints : Size=8
    Key : Lower : Body  : Upper
      7 :  None :  71.5 :  72.0
      8 :  None : 63.25 :  89.0


 dicision variable are real number
 objective 121.84615300000002

 Drone Model SHREY 11.076923

 Drone Model DHARAN 0.0

 Drone Model SATVA 0.0

 Drone Model ADHYA 0.0

 Drone Model PRAVAT 0.0
 dicision variable are integer
  objective 121.0

 Drone Model SHREY 11.0

 Drone Model DHARAN 0.0

 Drone Model SATVA 0.0

 Drone Model ADHYA 0.0

 Drone Model PRAVAT 0.0

10--
we take dicision variable are real number
6.5x0+8x1+10.5x2+11.75x3+14.5x4<=70
5.75x0+4.5x1+3.75x2+4.85x3+3.45x4<=90

In [321]:
model.constraints[7].deactivate()
model.constraints[8].deactivate()

In [322]:
for i in col_indices:
  model.x[i].domain=Reals

In [323]:
model.constraints.add(expr=6.5* model.x[0] +8* model.x[1] +10.5* model.x[2]+11.75*model.x[3]+14.5*model.x[4]<= 70)
model.constraints.add(expr= 5.75*model.x[0] +4.5* model.x[1] + 3.75* model.x[2] +4.85*model.x[3]+3.45*model.x[4]<= 90 )



In [324]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  11.0 :  None : False : False :  Reals
          1 :     0 :   0.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
  

In [325]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('solver termination codition:',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 118.4615385
  Upper bound: 118.4615385
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 6
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.029134035110473633
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
solver termination codition: optimal


In [326]:
print('\objective',model.objective())
for i in col_indices:
  print('\n Drone Model',data_csvfile['Drone Model'][i],model.x[i].value)
print('constraints')
model.constraints.display()

\objective 118.461541

 Drone Model SHREY 10.769231

 Drone Model DHARAN 0.0

 Drone Model SATVA 0.0

 Drone Model ADHYA 0.0

 Drone Model PRAVAT 0.0
constraints
constraints : Size=10
    Key : Lower : Body               : Upper
      9 :  None :         70.0000015 :  70.0
     10 :  None : 61.923078249999996 :  90.0


dicision variable are non negetive interger

In [327]:
for i in col_indices:
  model.x[i].domain=NonNegativeIntegers

In [328]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.769231 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True 

In [329]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('solver termination codition:',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 116.0
  Upper bound: 116.0
  Number of objectives: 1
  Number of constraints: 2
  Number of variables: 5
  Number of binary variables: 0
  Number of integer variables: 5
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.01
  Wallclock time: 0.01
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 16
  Error rc: 0
  Time: 0.03168463706970215
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
solver termination codition: optimal


In [330]:
print('\objective',model.objective())
for i in col_indices:
  print('\n Drone Model',data_csvfile['Drone Model'][i],model.x[i].value)
print('constraints')
model.constraints.display()

\objective 116.0

 Drone Model SHREY 7.0

 Drone Model DHARAN 3.0

 Drone Model SATVA 0.0

 Drone Model ADHYA 0.0

 Drone Model PRAVAT 0.0
constraints
constraints : Size=10
    Key : Lower : Body  : Upper
      9 :  None :  69.5 :  70.0
     10 :  None : 53.75 :  90.0


dicision variable are real number
objective 118.461541

 Drone Model SHREY 10.769231

 Drone Model DHARAN 0.0

 Drone Model SATVA 0.0

 Drone Model ADHYA 0.0

 Drone Model PRAVAT 0.0
 # dicision variable are integer
 objective 116.0

 Drone Model SHREY 7.0

 Drone Model DHARAN 3.0

 Drone Model SATVA 0.0

 Drone Model ADHYA 0.0

 Drone Model PRAVAT 0.0


10--
frist dicision variable are real number
6.5x0+8x1+10.5x2+11.75x3+14.5x4<=68
5.75x0+4.5x1+3.75x2+4.85x3+3.45x4<=91

In [331]:
model.constraints[9].deactivate()
model.constraints[10].deactivate()


In [332]:
for i in col_indices:
  model.x[i].domain=Reals

In [333]:
model.constraints.add(expr=6.5* model.x[0] +8* model.x[1] +10.5* model.x[2]+11.75*model.x[3]+14.5*model.x[4]<= 68)
model.constraints.add(expr= 5.75*model.x[0] +4.5* model.x[1] + 3.75* model.x[2] +4.85*model.x[3]+3.45*model.x[4]<= 91 )



In [334]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   7.0 :  None : False : False :  Reals
          1 :     0 :   3.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declara

In [335]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('termination condition:',result.solver.termination_condition)



Problem: 
- Name: unknown
  Lower bound: 115.0769231
  Upper bound: 115.0769231
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 6
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.030309438705444336
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
termination condition: optimal


In [336]:
print('\nobjective',model.objective())
for i in col_indices:
  print('Drone model',data_csvfile['Drone Model'][i],model.x[i].value)
print('\constraint')
model.constraints.display()


objective 115.07691799999999
Drone model SHREY 10.461538
Drone model DHARAN 0.0
Drone model SATVA 0.0
Drone model ADHYA 0.0
Drone model PRAVAT 0.0
\constraint
constraints : Size=12
    Key : Lower : Body               : Upper
     11 :  None :          67.999997 :  68.0
     12 :  None : 60.153843499999994 :  91.0


dicision variable are real number 
objective 115.07691799999999
Drone model SHREY 10.461538
Drone model DHARAN 0.0
Drone model SATVA 0.0
Drone model ADHYA 0.0
Drone model PRAVAT 0.0

In [337]:
for i in col_indices:
  model.x[i].domain=NonNegativeIntegers

In [338]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.461538 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :

In [339]:
result=opt_cbc.solve(model)
print(result)
print('solver status:',result.solver.status)
print('termination condition:',result.solver.termination_condition)



Problem: 
- Name: unknown
  Lower bound: 114.0
  Upper bound: 114.0
  Number of objectives: 1
  Number of constraints: 2
  Number of variables: 5
  Number of binary variables: 0
  Number of integer variables: 5
  Number of nonzeros: 5
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 1
  Error rc: 0
  Time: 0.027512788772583008
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

solver status: ok
termination condition: optimal


In [340]:
print('\nobjective',model.objective())
for i in col_indices:
  print('Drone model',data_csvfile['Drone Model'][i],model.x[i].value)
print('constraint')
model.constraints.display()


objective 114.0
Drone model SHREY 8.0
Drone model DHARAN 2.0
Drone model SATVA 0.0
Drone model ADHYA 0.0
Drone model PRAVAT 0.0
constraint
constraints : Size=12
    Key : Lower : Body : Upper
     11 :  None : 68.0 :  68.0
     12 :  None : 55.0 :  91.0


dicision variable are real number 
objective 115.07691799999999
Drone model SHREY 10.461538
Drone model DHARAN 0.0
Drone model SATVA 0.0
Drone model ADHYA 0.0
Drone model PRAVAT 0.0
#dicision variable are non negetive integer
objective 114.0
Drone model SHREY 8.0
Drone model DHARAN 2.0
Drone model SATVA 0.0
Drone model ADHYA 0.0
Drone model PRAVAT 0.0

frist constraints( Cost (in Crores of rupees) )are allways active and 2nd constraints(Parking space (sq. m.)) are inactive,  cost is decreasing then objective (carrying Capacity (kgs) )are  also decreasing ,